### ISSUE TITLE
Compiler warning in sillyrecords.cc

### ISSUE THREAD

**Issue by pieterlexis** [PowerDNS/pdns#5366]
- Program: Authoritative, Recursor,
 - Issue type: Bug report

### Short description

GCC 7.1 (`gcc (GCC) 7.1.1 20170516`) issues the following warning when compiling the recursor and the authoritative server:

```
sillyrecords.cc: In member function ‘virtual std::__cxx11::string LOCRecordContent::getZoneRepresentation(bool) const’:
sillyrecords.cc:296:8: warning: ‘%.2f’ directive output may be truncated writing between 4 and 313 bytes into a region of size between 0 and 37 [-Wformat-truncation=]
 string LOCRecordContent::getZoneRepresentation(bool noDot) const
        ^~~~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:939:0,
                 from /usr/include/c++/7.1.1/cstdio:42,
                 from /usr/include/c++/7.1.1/ext/string_conversions.h:43,
                 from /usr/include/c++/7.1.1/bits/basic_string.h:6159,
                 from /usr/include/c++/7.1.1/string:52,
                 from utility.hh:45,
                 from sillyrecords.cc:4:
/usr/include/bits/stdio2.h:65:44: note: ‘__builtin___snprintf_chk’ output between 48 and 1938 bytes into a destination of size 79
        __bos (__s), __fmt, __va_arg_pack ());
                                            ^
```

### Environment
 - Operating system: Arch Linux
 - Software version: master
 - Software source: git repo

### Steps to reproduce
1. `./bootstrap && ./configure --without-modules --without-dynmodules && make -C pdns sillyrecords.o`

### Expected behaviour
No warning 😄 

### Actual behaviour
The above warning is shown

**Comment by Habbie** [comment]
While poking at this I noticed we 32-bit truncate SIZE-encoded things. This helps:
```diff
diff --git a/pdns/sillyrecords.cc b/pdns/sillyrecords.cc
index 289cdb18c..970b2aa6f 100644
--- a/pdns/sillyrecords.cc
+++ b/pdns/sillyrecords.cc
@@ -16,7 +16,7 @@ static unsigned int poweroften[10] = {1, 10, 100, 1000, 10000, 100000,
 /* converts ascii size/precision X * 10**Y(cm) to 0xXY. moves pointer.*/
 static uint8_t precsize_aton(const char **strptr)
 {
-  unsigned int mval = 0, cmval = 0;
+  uint64_t mval = 0, cmval = 0;
   uint8_t retval = 0;
   const char *cp;
   int exponent;
```

I did not test much.

**Comment by dwfreed** [comment]
The 'f' format specifier expects a double, and if '%.2f' is given -DBL_MAX, it'll output 313 bytes.

**Comment by rgacogne** [comment]
I decided to take a good look at this warning, but it turns out that I can't reproduce it with either gcc 8.2.1 or clang 7.0.0. 
Still at first sight it looks like a false positive due to the limited range of values we are printing, but we could add some checks (either to the values, or check the return value of `snprintf()` and throw an exception if a truncation occurred) to be safe.

**Comment by dwfreed** [comment]
Fixed by 0e12635 in #7289
